x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/indexlifecycle/action/TransportMoveToStepAction.java

```diff
@@ -62,6 +62,16 @@ public ClusterState execute(ClusterState currentState) {
                         request.getNextStepKey());
                 }
 
+                @Override
+                public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                    IndexMetaData newIndexMetaData = newState.metaData().index(indexMetaData.getIndex());
+                    if (newIndexMetaData == null) {
```

Comment:
can this occur due to batching of updates?

Reply:
After talking with @DaveCTurner it seems like we won't have any batching here because batching occurs within the same instance of `ClusterStateTaskExecutor` and we don't implement batching ourselves here. However I think this check is still nice to have in case there are other factors at play.

Reply:
Can this be `assert newIndexMetaData != null`?

Reply:
I like David's suggestion here, as it suggests this should never happen more strongly than an if

Reply:
An assert is less strong though? because the check will not be done in production code, only in tests.

If we can envisage any scenarios where the `newState` passed to this method can be different to the state we returned in `execute()` then I think this whould stay as an `if` statement so we don't end up in a situation where we have a NPE thrown here because the index was deleted. IF we are confident that this kind of scenario should never occur and assert is fine.

Reply:
right. I guess I don't see this hurting, so I won't block it, but it may be misleading to people new to the code to walk through state in which this may be possible.

Reply:
We could add something like `assert false : "there should be no opportunity for the index to be deleted"` inside the if - that way we can catch it in testing while still handling it in production if there's a case we missed.  Does that sound reasonable, or is it too messy?

Reply:
Lets add an `assert newIndexMetaData != null` for testing but leave the if statement for production so we don't NPE if this ever does happen.

Reply:
I think we should leave this with the if statement so we are protected against NPEs. If we also want to add an assert to catch things in tests then that fine but I think the protection against a NPE in production should remain